aws/cloudformation/s3_backed_subdomain.yml

```diff
@@ -0,0 +1,112 @@
+---
+AWSTemplateFormatVersion: 2010-09-09
+Description:
+  "Basic template for provisioning subdomains which serve content from S3 via
+  CloudFront, like we do for the TTS audio files. We need three distinct
+  components for this functionality.
+ 
+  1. Route53 record mapping subdomain to CloudFront distribution (ie, tts.code.org)
+  2. CloudFront distribution with the domain of the S3 Bucket as origin (ie, cdo-tts.s3.amazonaws.com)
+  3. S3 Bucket with a BucketPolicy which grants global read access (ie, cdo-tts)
+ 
+  For more, see:
+ 
+  https://docs.aws.amazon.com/AmazonS3/latest/userguide/website-hosting-custom-domain-walkthrough.html
+  https://github.com/aws-samples/amazon-cloudfront-secure-static-site"
+
+Parameters:
+  BucketName:
+    Type: String
+    Default: cdo-dev-s3-backed-subdomain
+  CertificateArn:
+    Type: String
+  Domain:
+    Type: String
+    Default: dev-code.org
+  Subdomain:
+    Type: String
+    Default: s3-backed-subdomain
+
+Resources:
+  Bucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      AccessControl: PublicRead
+      BucketName: !Ref BucketName
+      Tags:
+        - Key: environment
+          Value: production
+
+  BucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref Bucket
+      PolicyDocument:
+        Statement:
+          - Effect: Allow
+            Principal: '*'
+            Action: 's3:GetObject'
+            Resource: !Join
+              - ''
+              - - 'arn:aws:s3:::'
+                - !Ref Bucket
+                - /*
+      Tags:
+        - Key: environment
+          Value: production
+
+  BucketBackedDistribution:
+    Type: AWS::CloudFront::Distribution
+    Properties:
+      DistributionConfig:
+        Enabled: true
+        DefaultCacheBehavior:
+          # Use the Managed-CachingOptimized policy from
+          # https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-cache-policies.html
+          CachePolicyId: 658327ea-f89d-4fab-a63d-7e88639e58f6
+          TargetOriginId: !Join
+              - '-'
+              - - 's3'
+                - !Ref BucketName
+          ViewerProtocolPolicy: https-only
+        Origins:
+          - DomainName: !Join
+              - '.'
+              - - !Ref BucketName
+                - 's3.amazonaws.com'
+            Id: !Join
+              - '-'
+              - - 's3'
+                - !Ref BucketName
+            S3OriginConfig:
+              OriginAccessIdentity: ""
+        ViewerCertificate:
+          AcmCertificateArn: !Ref CertificateArn
+          MinimumProtocolVersion: TLSv1
+          SslSupportMethod: sni-only
+      Tags:
+        - Key: environment
+          Value: production
+
+  SubdomainRecord:
+    Type: AWS::Route53::RecordSet
+    Properties:
+      HostedZoneName: !Join
+        - ''
+        - - !Ref Domain
+          - '.'
+      Name: !Join
```

Comment:
I should have brought this up earlier. Not critical since the current syntax has been tested in our Dev account, but using `!Sub` may be more readable. This blog has some good comparisons between !Join and !Sub illustrating the trade offs.

Reply:
oh, nice! I was looking for string interpolation and couldn't find it. Probably because they for some reason decided to call it `!Sub` :upside_down_face: